errors/errors.go

```diff
@@ -97,8 +103,15 @@ func Register(code uint32, description string) *Error {
 
 // usedCodes is keeping track of used codes to ensure their uniqueness. No two
 // error instances should share the same error code.
-var usedCodes = map[uint32]*Error{
-	1: nil, // Error code 1 is restricted for non-weave errors and must not be used.
+var usedCodes = map[uint32]*Error{}
+
+// Lookup will resolve the abci code to a root Error object
+// if nothing is registered for that code, return errInternal
+func Lookup(code uint32) *Error {
```

Comment:
If the only use for this is the [weave client](https://github.com/iov-one/weave/pull/549/files#diff-1cd41a2611da49ce63630d5a51bd82c5R53) then I think this could be adjusted.

How about changing this function to
```go
// ABCIError is a work in progress name. It cannot be called Lookup anymore.
func ABCIError(code uint32, description string) *Error {
    if e, ok := usedCodes[code]; ok {
        return errors.Wrap(e, description)
    } 
    return &Error{code: code, description: description}
}
```

My undestanding of the functionality is that you are using a client and want to test the error. For example

```go
_, err := client.DoSomething()
if errors.ErrNotFound.Is(err) {
   // do something
}
```

If the returned by the API error is not registerd in `usedCodes` then this whole pattern is useless. That is ok because if the error is not registered it does not exist as a variable as well. The `Is` method cannot be called on something that does not exist.

I think creating an error on the fly when it is not registered is good compromise.

In case of the original proposal, this mistake can happen:
```go
_, err1 := client.DoSomething() // Returns unregistered error code 66666
_, err2 := client.DoSomethingElse() // Returned unregistered error code 9999
err1.Is(err2) // true, because all unknown errors are generalized to a single instance
```

In case of ad-hoc created errors, there is another issue:
```go
_, err1 := client.DoSomething() // Returns unregistered error code 6666
_, err2 := client.DoSomething() // Returns unregistered error code 6666
err1.Is(err2) // false because unknown errors are created on the fly and instances are never equal
```

I think preserving the orginal error code instead of downgrading to 1 migth come handy when debugging as well. In the end that is what those errors are for in the first place.

What do you think?

Reply:
Good suggestion. I will do just that.